examples/cosmos/ics20/ics20.qnt

```diff
@@ -0,0 +1,242 @@
+  /// the channel `C` connects to, in the other chain.
+  pure def getCounterparty(chainState: ChainState, sourceChannel: Channel): Channel = {
+    chainState.channels.get(sourceChannel)
+  }
+
+  pure def sendFungibleTokens(chainState: ChainState, denomination: DenomTrace, amount: UINT256,
+                              sender: Address, receiver: Address, sourcePort: str, sourceChannel: Channel,
+                              timeoutHeight: Height, timeoutTimestamp: uint64): ChainState = {
+    val bankResult = if (movingBackAlongTrace({port: sourcePort, channel: sourceChannel}, denomination)) {
+      // burn vouchers
+      (chainState.bank).BurnCoins(sender, denomination, amount)
+    } else {
+      // escrow tokens
+      val escrowAccount = chainState.channelEscrowAddresses.get(sourceChannel)
+      chainState.bank.TransferCoins(sender, escrowAccount, denomination, amount)
+    }
+
+    if (bankResult.success) {
+      val data = { denom: denomination, amount: amount, sender: sender, receiver: receiver, memo: "" }
+
+      // handler.sendPacket
+      val packet = {
+        data: data,
+        sourcePort: sourcePort,
+        sourceChannel: sourceChannel,
+        destPort: "transfer",
+        destChannel: getCounterparty(chainState, sourceChannel)
+      }
+
+      chainState
+        .with("bank", bankResult.accounts)
+        .with("outPackets", chainState.outPackets.union(Set(packet)))
+    } else {
+      chainState
+    }
+  }
+
+  pure def onRecvPacket(chainState: ChainState, packet: Packet): ReceiveTokenPacketResult = {
+    val data = packet.data
+    if (movingBackAlongTrace({port: packet.sourcePort, channel: packet.sourceChannel}, data.denom)) {
+      // unescrow tokens to receiver
+      val escrowAccount = chainState.channelEscrowAddresses.get(packet.destChannel)
+      val receiverDenom = updateTrace(data.denom, packet.sourcePort, packet.sourceChannel, packet.destPort, packet.destChannel)
+
+      val bankResult = chainState.bank.TransferCoins(escrowAccount, data.receiver, receiverDenom, data.amount)
+
+      val newChainState = chainState.with("bank", bankResult.accounts)
+      val ack = if (bankResult.success) {
+        { success: true, errorMessage: "" }
+      } else {
+        { success: false, errorMessage: "transfer coins failed" }
+      }
+
+      { chainState: newChainState, acknowledgement: ack }
+    } else {
+      // mint vouchers to receiver
+      val newDenom = updateTrace(data.denom, packet.sourcePort, packet.sourceChannel, packet.destPort, packet.destChannel)
+
+      val bankResult = chainState.bank.MintCoins(data.receiver, newDenom, data.amount)
+
+      val newChainState = chainState.with("bank", bankResult.accounts)
+      val ack = if (bankResult.success) {
+        { success: true, errorMessage: "" }
+      } else {
+        { success: false, errorMessage: "mint coins failed" }
+      }
+
+      { chainState: newChainState, acknowledgement: ack }
```

Comment:
This logic is the same as in the `if`-branch, except for the `errorMessage` literal.

I'm wondering if we should fold de-duplicate this by moving this chunk outside the if-then-else. We can disambiguate the error message based on the result of `movingBackAlongTrace`.

Reply:
Done!